crates/asm/subprotocols/checkpoint/src/state.rs

```diff
@@ -0,0 +1,146 @@
+        &self.checkpoint_predicate
+    }
+
+    /// Returns the verified epoch summary, if any.
+    pub fn verified_epoch_summary(&self) -> Option<&EpochSummary> {
+        self.verified_epoch_summary.as_ref()
+    }
+
+    /// Get the expected next epoch number.
+    pub fn expected_next_epoch(&self) -> Epoch {
+        self.verified_epoch_summary
+            .as_ref()
+            .map(|s| s.epoch() + 1)
+            .unwrap_or(0)
+    }
+
+    /// Update the sequencer credential.
+    pub fn update_sequencer_cred(&mut self, new_key: Buf32) {
+        self.sequencer_cred = CredRule::SchnorrKey(new_key);
+    }
+
+    /// Update the checkpoint predicate.
+    pub fn update_checkpoint_predicate(&mut self, new_predicate: PredicateKey) {
+        self.checkpoint_predicate = new_predicate;
+    }
+
+    /// Returns the height of the last L1 block covered by the previous checkpoint.
+    pub fn last_covered_l1_height(&self) -> u32 {
+        self.last_covered_l1.height
+    }
+
+    /// Returns the last L1 block commitment covered by the previous checkpoint.
+    pub fn last_covered_l1(&self) -> L1Commitment {
+        self.last_covered_l1
+    }
+
+    /// Returns the slot of the last L2 terminal block.
+    /// Returns `None` before the first checkpoint is verified.
+    pub fn last_l2_terminal_slot(&self) -> Option<u64> {
+        self.verified_epoch_summary
+            .as_ref()
+            .map(|s| s.terminal().slot())
+    }
+
+    /// Returns the last L2 terminal block commitment, if any.
+    pub fn last_l2_terminal(&self) -> Option<OLBlockCommitment> {
+        self.verified_epoch_summary
+            .as_ref()
+            .map(|s| *s.terminal())
+    }
+
+    /// Returns the pre-state root for the next checkpoint.
+    ///
+    /// This is the final state from the last verified epoch, or zero for the first checkpoint.
+    pub fn pre_state_root(&self) -> Buf32 {
+        self.verified_epoch_summary
+            .as_ref()
+            .map(|s| *s.final_state())
+            .unwrap_or_else(Buf32::zero)
+    }
+
+    /// Update state with a verified checkpoint.
+    ///
+    /// Called after the checkpoint signature and proof have been verified.
+    pub fn update_with_checkpoint(&mut self, checkpoint: &CheckpointPayload) {
+        let batch_info = &checkpoint.commitment.batch_info;
+        let transition = &checkpoint.commitment.transition;
+
+        // prev_terminal comes from current state (the terminal of the previous epoch).
+        // For the first checkpoint (epoch 0), this will be null/zero.
+        let prev_terminal = self
+            .verified_epoch_summary
+            .as_ref()
+            .map(|s| *s.terminal())
+            .unwrap_or_else(OLBlockCommitment::null);
+
+        let epoch_summary = EpochSummary::new(
+            batch_info.epoch,
+            batch_info.l2_range.end, // terminal: this epoch's final L2 block
+            prev_terminal,           // prev_terminal: from current state
```

Comment:
as above, if this is the state, let's just call it state (note also the prev/current conflict you have here)